Chapter 8 (#8)
Andarist/total-typescript-book: book-content/chapters/08-classes.md

Author: I feel like this chapter is missing:
- `typeof Cls` vs `Cls` (the first one is the type of the class - the second one is the type of an instance of this class)
- that TS is structural in nature and that, unlike many other languages, TS doesn't have nominal classes... unless u have private properties. I think this is a big gotchat for newcomers, especially for those coming from other backgrounds

You could also mention gotchas around bivariant methods here but you didn't get to related topics just yet so perhaps it's not the right time yet. Although methods bivariance is hugely motivated by classes

class Album {
  title: string;
  artist: string;
  releaseYear: number;

  constructor() {
    this.title = "Loop Finding Jazz Records";
    this.artist = "Jan Jelinek";
    this.releaseYear = 2001;
  }
}
```

Now, when we create a new instance of the `Album` class, we can access the properties and values we've set in the constructor:

```typescript
const loopFindingJazzRecords = new Album();

console.log(loopFindingJazzRecords.title); // Output: Loop Finding Jazz Records
```

The `new` keyword creates a new instance of the `Album` class, and the constructor sets the initial values of our class's properties. In this case, because the properties are hardcoded, every instance of the `Album` class will have the same values.

#### You Don't Always Need To Type Class Properties

As we'll see, TypeScript can do some really smart inference with classes. It's able to infer the types of the properties from where we assign them in the constructor, so we can actually drop some of the type annotations:

```typescript
class Album {
  title;
  artist;
  releaseYear;

  constructor() {
    this.title = "Loop Finding Jazz Records";
    this.artist = "Jan Jelinek";
    this.releaseYear = 2001;
  }

> }

Author: It's simplifiable even further, as long as we don't need to depend on any arguments:
```ts
class Album {
  title = "Loop Finding Jazz Records";
  artist = "Jan Jelinek";
  releaseYear = 2001;
}
```

```

However, it's common to see the types specified in the class body as well since they act as a form of documentation for the class that's quick to read.

### Adding Arguments to the Constructor

We can use the constructor to declare arguments for the class. This allows us to pass in values when creating a new instance of the class.

Update the constructor to accept an `opts` argument that includes the properties of the `Album` class:

```typescript
// inside the Album class
constructor(opts: { title: string; artist: string; releaseYear: number }) {
 // ...
}
```

Then inside of the body of the constructor, we'll use assign `this.title`, `this.artist`, and `this.releaseYear` to the values of the `opts` argument.

```typescript
// inside the Album class
constructor(opts: { title: string; artist: string; releaseYear: number }) {
  this.title = opts.title;
  this.artist = opts.artist;
  this.releaseYear = opts.releaseYear;
}
```

The `this` keyword refers to the instance of the class, and it's used to access the properties and methods of the class.

> Now, when we create a new instance of the `Album` class, we can pass an object with the properties we want to set. If we don't provide any values, the default values will be used:

Author: you mention default values here and u even rely on them in the latter example but u have no default values here anywhere - u only introduce them later which is pretty confusing at this moment of reading

```typescript
const loopFindingJazzRecords = new Album({
  title: "Loop Finding Jazz Records",
  artist: "Jan Jelinek",
  releaseYear: 2001,
});

console.log(loopFindingJazzRecords.title); // Output: Loop Finding Jazz Records

const unknownAlbum = new Album();

console.log(unknownAlbum.title); // Output: Unknown Album
```

### Using a Class as a Type

An interesting property of classes in TypeScript is that they can be used as types for variables and function parameters. The syntax is similar to how you would use any other type or interface.

In this case, we'll use the `Album` class to type the `album` parameter of a `printAlbumInfo` function:

```typescript
function printAlbumInfo(album: Album) {
  console.log(
    `${album.title} by ${album.artist}, released in ${album.releaseYear}.`,
  );
}
```

We can then call the function and pass in an instance of the `Album` class:

```typescript
printAlbumInfo(sixtyNineLoveSongsAlbum);

// Output: 69 Love Songs by The Magnetic Fields, released in 1999.
```

> While using a class as a type is possible, it's a much more common pattern to require classes to implement a specific interface.

Author: this feels pretty opinionated, in many codebases I've seen classes were used as types directly - if you don't share an interface between multiple classes then why would I bother redeclaring every property/method to get that interface?

## Properties in Classes

Now that we've seen how to create a class and create new instances of it, let's look a bit closer at how properties work.

### Class Property Initializers

You can set default values for properties directly in the class body. These are called class property initializers.

```typescript
class Album {
  title = "Unknown Album";
  artist = "Unknown Artist";
  releaseYear = 0;
}
```

You can combine them with type annotations:

```typescript
class Album {
  title: string = "Unknown Album";
  artist: string = "Unknown Artist";
  releaseYear: number = 0;
}
```

Importantly, class property initializers are resolved _before_ the constructor is called. This means you can override the default values by assigning a different value in the constructor:

```typescript
class User {
  name = "Unknown User";

  constructor() {
    this.name = "Matt Pocock";
  }
}

const user = new User();

console.log(user.name); // Output: Matt Pocock
```

### `readonly` Class Properties

> As we've seen with types and interfaces, the `readonly` keyword can be used to make a property immutable. This means that once the property is set, it cannot be changed:

Author: ```suggestion
As we've seen with types and interfaces, the `readonly` keyword can also be used with class properties. This means that once the property is set, it cannot be changed:
```

```typescript
class Album {
  readonly title: string;
  readonly artist: string;
  readonly releaseYear: number;
}
```

### Optional Class Properties

We can also mark properties as optional in the same way as objects, using the `?:` annotation:

```typescript
class Album {
  title?: string;
  artist?: string;
  releaseYear?: number;
}
```

As we can see from the lack of errors above, this also means they don't need to be set in the constructor.

### `public` and `private` properties

> The `public` and `private` keywords are used to control the visibility and accessibility of class properties.

Author: calling out both visibility and accessibility here calls out for defining both (or at least for differentiating between both). It's kinda the same thing here so I think I'd avoid using both terms and stick to one

  releaseYear: number;

  constructor(opts: { title: string; artist: string; releaseYear: number }) {
    this.title = title;
    this.artist = artist;
    this.releaseYear = releaseYear;
  }

  displayInfo() {
    console.log(
      `${this.title} by ${this.artist}, released in ${this.releaseYear}.`,
    );
  }
}
```

The goal is to create a `SpecialEditionAlbum` class that extends the `Album` class and adds a `bonusTracks` property.

### Extending a Class

The first step is to use the `extends` keyword to create the `SpecialEditionAlbum` class:

```typescript
class SpecialEditionAlbum extends Album {}
```

Once the `extends` keyword is added, any new properties or methods added to the `SpecialEditionAlbum` class will be in addition to what it inherits from the `Album` class. For example, we can add a `bonusTracks` property to the `SpecialEditionAlbum` class:

```typescript
class SpecialEditionAlbum extends Album {
  bonusTracks: string[];
}
```

Next, we need to add a constructor that includes all of the properties from the `Album` class as well as the `bonusTracks` property. There are a couple of important things to note about the constructor when extending a class.

> First, the arguments to the constructor should match the shape used in the parent class. In this case, that's an `opts` object with the properties of the `Album` class along with the new `bonusTracks` property.

Author: There is no such requirement, I wouldn't recommend deviating too far but this is perfectly valid code for TS:
```ts
class Base1 {
  constructor(arg: number) {}
}
class Cls1 extends Base1 {
  constructor(arg: string) {
    super(parseInt(arg));
  }
}

class Base2 {
  constructor(arg: { foo: string }) {}
}
class Cls2 extends Base2 {
  constructor(arg: { bar: string }) {
    super({ foo: arg.bar });
  }
}
```

Here we have a class called `CanvasNode` that currently functions identically to an empty object:

```typescript
class CanvasNode {}
```

Inside of a test case, we instantiate the class by calling `new CanvasNode()`.

However, have some errors since we expect it to house two properties, specifically `x` and `y`, each with a default value of `0`:

```ts twoslash
// @errors: 2339
import { it, expect } from "vitest";

class CanvasNode {}

// ---cut---
it("Should store some basic properties", () => {
  const canvasNode = new CanvasNode();

  expect(canvasNode.x).toEqual(0);
  expect(canvasNode.y).toEqual(0);

  // @ts-expect-error Property is readonly
  canvasNode.x = 10;

  // @ts-expect-error Property is readonly
  canvasNode.y = 20;
});
```

As seen from the `@ts-expect-error` directives, we also expect these properties to be readonly.

Your challenge is to implement the `CanvasNode` class to satisfy these requirements. For extra practice, solve the challenge with and without the use of a constructor.

<Exercise title="Exercise 1: Creating a Class" filePath="/src/030-classes/108-understand-classes.problem.ts"></Exercise>

### Exercise 2: Implementing Class Methods

> In this exercise, we've simplified our `CanvasNode` class so that it no longer has read-only properties:

Author: read-only vs readonly spelling is not used consistently across the book

    this.x = position?.x ?? 0;
    this.y = position?.y ?? 0;
  }

  move(x: number, y: number) {
    this.x = x;
    this.y = y;
  }
}

// ---cut---
it("Should be able to move", () => {
  const canvasNode = new CanvasNode();

  expect(canvasNode.position).toEqual({ x: 0, y: 0 });

  canvasNode.move(10, 20);

  expect(canvasNode.position).toEqual({ x: 10, y: 20 });
});

it("Should be able to receive an initial position", () => {
  const canvasNode = new CanvasNode({
    x: 10,
    y: 20,
  });

  expect(canvasNode.position).toEqual({ x: 10, y: 20 });
});
```

Your task is to update the `CanvasNode` class to include a `position` getter that will allow for the test cases to pass.

<Exercise title="Exercise 3: Implement a Getter" filePath="/src/030-classes/111-getters.problem.ts"></Exercise>

### Exercise 4: Implement a Setter

The `CanvasNode` class has been updated so that `x` and `y` are now private properties:

```typescript
class CanvasNode {
  #x: number;
  #y: number;

  constructor(position?: { x: number; y: number }) {
    this.#x = position?.x ?? 0;
    this.#y = position?.y ?? 0;
  }

  // your `position` getter method here

  // move method as before
}
```

> The `#` in front of the `x` and `y` properties means they are `readonly` and can't be modified directly outside of the class. In addition, when a getter is present without a setter, its property will also be treated as `readonly`, as seen in this test case:

Author: ```suggestion
The `#` in front of the `x` and `y` properties means they are private and can't be modified directly outside of the class. In addition, when a getter is present without a setter, its property will also be treated as `readonly`, as seen in this test case:
```

In general, something weird is going on here and in subsequent paragraphs. You mention `readonly`, and getters without setters but there are no getters here. There is a mismatch between text and the presented code.

Author: I know that getter was supposed to be introduced by the reader in the previous exercise but it's still confusing to me. You didn't even introduce getters and setters at all before, and now you softly expect the reader to be aware of them.

Those exercises are about classes but now you present a simple object type with a simple property as a validation for this statement:
> In addition, when a getter is present without a setter, its property will also be treated as `readonly`, as seen in this test case:

I don't feel like the example is related to this paragraph. That example has a readonly property because it has a readonly modifier on it - not because it's a getter-only property.

If you like, you can use an `abstract` class to define `Shape`.

<Exercise title="Exercise 5: Extending a Class" filePath="/src/030-classes/114-extending-other-classes.problem.ts"></Exercise>

### Solution 1: Creating a Class

Here's an example of a `CanvasNode` class with a constructor that meets the requirements:

```typescript
class CanvasNode {
  readonly x: number;
  readonly y: number;

  constructor() {
    this.x = 0;
    this.y = 0;
  }
}
```

Without a constructor, the `CanvasNode` class can be implemented by assigning the properties directly:

```typescript
class CanvasNode {
  readonly x = 0;
  readonly y = 0;
}
```

### Solution 2: Implementing Class Methods

The `move` method can be implemented either as a regular method or as an arrow function:

Here's the regular method:

```typescript
class CanvasNode {
  x = 0;
  y = 0;

  move(x: number, y: number) {
    this.x = x;
    this.y = y;
  }
}
```

And the arrow function:

```typescript
class CanvasNode {
  x = 0;
  y = 0;

  move = (x: number, y: number) => {
    this.x = x;
    this.y = y;
  };
}
```

> As discussed in a previous section, it's safer to use the arrow function to avoid issues with `this`.

Author: Safer but less memory-efficient:
```ts
class Test1 {
  method() {}
}

new Test1().method === new Test1().method; // true

class Test2 {
  method = () => {};
}

new Test2().method === new Test2().method; // false
```
